torch/csrc/jit/tensorexpr/codegen.cpp

```diff
@@ -83,21 +83,136 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
       false, "This codegen backend does not implement call_with_numel");
 }
 
-StmtPtr insertAllocFree(std::unordered_set<BufPtr>& interm_bufs, StmtPtr stmt) {
+c10::optional<size_t> bufSize(BufPtr buf) {
+  size_t size = elementSize(buf->dtype().scalar_type()) * buf->dtype().lanes();
+  for (auto& d : buf->dims()) {
+    if (!d->isConstant()) {
+      return c10::nullopt;
+    }
+    size = size * (*intValue(d));
+  }
+  return size;
+}
+
+// This algorithm takes the list of intermediate buffers and their liveness
+// ranges, and returns the allocations of these buffers. A buffer 'A' can be
+// allocated in the memory (appears as a pair of 'A's in the allocation results)
+// or reuse another buffer such as 'B' (appears as ('A', 'B')). Specifically, we
+// linearly scan the intermediate buffers by the time they appear, and try to
+// assign it an existing non-occupied memory allocation. If there are no such
+// allocations available, we'll create memory for it. Once we are beyond the
+// liveness range of this buffer, we'll mark its corresponding memory allocation
+// as "up for grabs" for future reuse.
+std::vector<std::pair<BufPtr, BufPtr>> AllocBufsWithMemReuse(
+    const std::unordered_set<BufPtr>& bufs,
+    const std::unordered_map<BufPtr, std::tuple<int32_t, int32_t>>&
+        buf_ranges) {
+  // Sort buffers by the time they appear.
+  std::vector<BufPtr> bufs_sorted(bufs.begin(), bufs.end());
+  auto sorting_function_by_start_time = [&buf_ranges](
+                                            BufPtr b1, BufPtr b2) -> bool {
+    return std::get<0>(buf_ranges.at(b1)) < std::get<0>(buf_ranges.at(b2));
+  };
+  std::sort(
+      bufs_sorted.begin(), bufs_sorted.end(), sorting_function_by_start_time);
+
+  // Map intermediate buffers to the most recently used memory if any.
+  std::list<BufPtr> mem_up_for_grabs;
+  std::unordered_map<BufPtr, BufPtr> buf_mem_map;
+  std::vector<std::pair<BufPtr, BufPtr>> buf_allocs;
+
+  auto sorting_function_by_end_time = [&buf_ranges](
+                                          BufPtr b1, BufPtr b2) -> bool {
+    return std::get<1>(buf_ranges.at(b1)) < std::get<1>(buf_ranges.at(b2));
+  };
+  std::vector<BufPtr> buf_to_release;
+  for (auto buf : bufs_sorted) {
+    auto start = std::get<0>(buf_ranges.at(buf));
+    auto end = std::get<1>(buf_ranges.at(buf));
+
+    // Release memory for buffers whose liveness range ends before the creation
+    // time of this buf.
+    // TODO: optimize in-place opererations and copy operations
+    buf_to_release.clear();
```

Comment:
Nit: instead of clearing we could just declare `buf_to_release` here - this would make it more obvious that its value is only needed within a single iteration of this loop.